// @crystallize/js-storefrontaware-utils/dist/storefront.server.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createStoreFront = void 0;
const js_api_client_1 = require("@crystallize/js-api-client");
/**
 * Create store TStoreFront
 * Load configuration from adapter and configure an apiClient for the store front
 *
 * @param adapter A TStoreFrontAdapter instance
 * @param options Object with options, or the deprecated syntax of a boolean for just `withSecrets`
 * @returns The loaded config + a configured api client
 */
const createStoreFront = async (adapter, options = false, clientOptions) => {
    options = typeof options === 'boolean' ? { withSecrets: options } : options;
    const withSecrets = options?.withSecrets ?? false;
    const config = await adapter.config(withSecrets);
    return {
        config,
        apiClient: (0, js_api_client_1.createClient)({
            tenantIdentifier: config.tenantIdentifier,
            tenantId: config.tenantId,
            accessTokenId: withSecrets ? config.configuration?.CRYSTALLIZE_ACCESS_TOKEN_ID || '' : '',
            accessTokenSecret: withSecrets ? config.configuration?.CRYSTALLIZE_ACCESS_TOKEN_SECRET || '' : '',
        }, clientOptions),
    };
};
exports.createStoreFront = createStoreFront;
